run_tests.py

```diff
@@ -80,39 +88,146 @@ def decompile_and_compare(
     return should_overwrite or not changed
 
 
-def decompile_and_capture_output(options: Options) -> str:
+def decompile_and_capture_output(options: Options, brief_crashes: bool) -> str:
     out_string = io.StringIO()
     with contextlib.redirect_stdout(out_string):
         returncode = decompile(options)
+    out_text = out_string.getvalue()
+    # Rewrite paths in the output to be relative (e.g. in tracebacks)
+    out_text = out_text.replace(str(Path(__file__).parent), ".")
     if returncode == 0:
-        return out_string.getvalue()
+        return out_text
     else:
-        return CRASH_STRING
+        if brief_crashes:
+            return CRASH_STRING
+        else:
+            return f"{CRASH_STRING}\n{out_text}"
 
 
 def run_e2e_test(
-    e2e_top_dir: Path, e2e_test_path: Path, should_overwrite: bool, coverage: Any
+    e2e_top_dir: Path,
+    e2e_test_path: Path,
+    should_overwrite: bool,
+    filter_regex: Optional[str],
+    coverage: Any,
 ) -> bool:
-    logging.info(f"Running test: {e2e_test_path.name}")
 
     ret = True
     for asm_file_path in e2e_test_path.glob("*.s"):
         old_output_path = asm_file_path.parent.joinpath(asm_file_path.stem + "-out.c")
         flags_path = asm_file_path.parent.joinpath(asm_file_path.stem + "-flags.txt")
+
+        name = f"e2e:{asm_file_path.relative_to(e2e_top_dir)}"
+        if filter_regex is not None and not re.search(filter_regex, name):
+            continue
         if coverage:
-            coverage.switch_context(str(asm_file_path.relative_to(e2e_top_dir)))
+            coverage.switch_context(name)
+        logging.info(f"Running test: {name}")
+
         if not decompile_and_compare(
-            asm_file_path, old_output_path, flags_path, should_overwrite
+            asm_file_path,
+            old_output_path,
+            brief_crashes=True,
+            should_overwrite=should_overwrite,
+            flags_path=flags_path,
+            flags=["test"],
         ):
             ret = False
     return ret
 
 
-def main(should_overwrite: bool, coverage: Any) -> int:
+def run_project_tests(
+    base_dir: Path,
+    output_dir: Path,
+    should_overwrite: bool,
+    filter_regex: Optional[str],
+    name_prefix: str,
+    coverage: Any,
+) -> bool:
+    ret = True
+    asm_dir = base_dir / "asm"
+    context_file = base_dir / "ctx.c"
+    for asm_file in asm_dir.rglob("*"):
+        if asm_file.suffix not in (".asm", ".s"):
+            continue
+        if "non_matching" in str(asm_file):
+            continue
+
+        asm_name = asm_file.name
+        if (
+            asm_name.startswith("code_data")
+            or asm_name.startswith("code_rodata")
+            or asm_name.startswith("boot_data")
+            or asm_name.startswith("boot_rodata")
+            or asm_name.endswith("_data.asm")
+            or asm_name.endswith("_rodata.asm")
+        ):
+            continue
+
+        has_context = context_file.exists()
```

Comment:
maybe worth basing this on a command line parameter since both behaviors are interesting to test and they use different expectation filenames

Reply:
Yeah you bring up a great point. I was trying to keep the arguments simple, but keep the program flexible enough to run with partial context. 

I'm going to try to refactor this a bit.